File: 0022-generate-parentheses/solution.cpp
class Solution {
public:

    void backtrack(int open,int close,vector<string>& ans,string s){
        if(open==0 && close==0){
            
            ans.push_back(s);
            
            return;
        }
        if(open>0){
            backtrack(open-1,close,ans,s+"(");
        }
        if(open<close){
            backtrack(open,close-1,ans,s+")");
        }

    }
    vector<string> generateParenthesis(int n) {
        vector<string> ans;
        if(n==0) return ans;
        backtrack(n,n,ans,"");
        return ans;
    }
};

File: 0022-generate-parentheses/solution.java
class Solution {
    void backtrack(List<String> ans,String s,int o,int c){
    if(o==0 && c==0){
        ans.add(s);
        return;
    }
    if(o>0) backtrack(ans,s+'(',o-1,c);
    if(o<c) backtrack(ans,s+')',o,c-1);
}
    public List<String> generateParenthesis(int n) {
        List<String> ans=new ArrayList<>();
        backtrack(ans,new String(),n,n);
        return ans;
    }
}
